util/throttle.cc

```diff
@@ -140,6 +142,13 @@ ThrottleThread(
             tot_compact+=gThrottleData[loop].m_Compactions;
         }   // for
 
+        // capture current state of level-0 and other levels' backlog
+        gThrottleData[replace_idx].m_Backlog=gCompactionThreads->m_WorkQueueAtomic;
+        gPerfCounters->Add(ePerfThrottleBacklog1, gThrottleData[replace_idx].m_Backlog);
+
+        gThrottleData[0].m_Backlog=gLevel0Threads->m_WorkQueueAtomic;
```

Comment:
gThrottleData[0].m_Backlog used to be a “live” variable.  This branch changes its definition.  It is now only set during this routine.  It is currently safe.

I would have a defensible position on gThrottleData[0].m_Keys and gThrottleData[0].m_Compactions being used unprotected in the else condition had they been declared as volatile.  But I have a vague memory of volatiles in a structure caused an issue on some compilers.  Maybe gcc wanted volatile at the struct level and some other compiler wanted it at the member variable level.  I just backed off.

This routine loops only once a minute and the mutex may get hit 60 to 120 times per second.  It is not a bottleneck.  Therefore moving discrete lock / unlock calls to encompass a larger portion of code.  These discrete calls can later be replaced by mutex objects.

> On Aug 9, 2015, at 10:16 PM, Paul A. Place <notifications@github.com> wrote:
> 
> In util/throttle.cc <https://github.com/basho/leveldb/pull/162#discussion_r36598057>:
> 
> > @@ -140,6 +142,13 @@ ThrottleThread(
> >              tot_compact+=gThrottleData[loop].m_Compactions;
> >          }   // for
> >  
> > +        // capture current state of level-0 and other levels' backlog
> > +        gThrottleData[replace_idx].m_Backlog=gCompactionThreads->m_WorkQueueAtomic;
> > +        gPerfCounters->Add(ePerfThrottleBacklog1, gThrottleData[replace_idx].m_Backlog);
> > +
> > +        gThrottleData[0].m_Backlog=gLevel0Threads->m_WorkQueueAtomic;
> gThrottleData[0] is accessed in several places outside the lock, which means we could have an inconsistent state for gThrottleData[0] if someone calls SetThrottleWriteRate() for level 0 while we're taking care of business (e.g., on line 179); before releasing the lock on line 126, should make a copy of gThrottleData[0] and clear it (moving the memset on line 214) and then access the copy outside the lock
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/basho/leveldb/pull/162/files#r36598057>.
>